src/index.ts

```diff
@@ -185,50 +201,102 @@ export function makeMemlet(disklet: Disklet): Memlet {
     // Add file to the store files map
     state.store.files[key] = file
+
+    // Add file's size to memory usage
+    await adjustMemoryUsage(file.size)
   }
 
   // Used to add undefined type checking to file retrieval
   function getStoreFile(key: string): File | undefined {
     return state.store.files[key]
   }
 
-  function deleteStoreFile(key: string): File | undefined {
+  async function deleteStoreFile(key: string): Promise<File | undefined> {
     const file = getStoreFile(key)
 
     if (file != null) {
-      adjustMemoryUsage(-file.size)
+      // Deleting file from store should invoke adjustMemoryUsage again
+      await adjustMemoryUsage(-file.size)
       // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
       delete state.store.files[key]
     }
 
     return file
   }
 
-  function adjustMemoryUsage(bytes?: number): void {
+  async function adjustMemoryUsage(bytes?: number): Promise<void> {
     if (bytes != null) {
       state.store.memoryUsage += bytes
     }
 
     // Remove files if memory usage exceeds maxMemoryUsage
     if (state.store.memoryUsage > state.config.maxMemoryUsage) {
-      const fileEntry = state.fileQueue.dequeue()
-      if (fileEntry != null) {
-        // Deleting file from store will invoke adjustMemoryUsage again
-        deleteStoreFile(fileEntry.key)
+      const file = state.writtenFileQueue.dequeue()
+
+      if (file != null) {
+        await deleteStoreFile(file.key)
+        return
+      }
+
+      const memoryOnlyFile = state.memoryOnlyFileQueue.dequeue()
+
+      if (memoryOnlyFile != null) {
+        await persistFile(memoryOnlyFile)
+        await deleteStoreFile(memoryOnlyFile.key)
       }
     }
   }
 
   function getCacheKey(path: string): string {
     return `${memletInstanceId}:${path}`
   }
+
+  async function persistFile(file: File): Promise<void> {
+    const path = fileKeyToPath(file.key)
+    const dataString = JSON.stringify(file.data)
+
+    await disklet.setText(path, dataString)
+  }
+
+  async function persistMemoryOnlyFiles(): Promise<void> {
+    for (let i = 0; i < MAX_BATCH_SIZE; ++i) {
+      // Pull out any memory-only files
+      const file = state.memoryOnlyFileQueue.dequeue()
+
+      // Exit loop if no memory-only files
+      if (file == null) break
+
+      // Persist file
+      await persistFile(file)
+
+      // Move file to written file queue
+      state.writtenFileQueue.enqueue(file)
+    }
+  }
+
+  /**
+   * Scheduler to drain file's from memory onto disk.
+   * This is run continually at a constant interval once invoked until
+   * there are no more memory-only files.
+   */
+  function drainMemoryOnlyFiles(): void {
+    setTimeout(() => {
+      persistMemoryOnlyFiles()
+        .then(() => {
+          if (state.memoryOnlyFileQueue.list().length > 0) {
+            drainMemoryOnlyFiles()
+          }
+        })
+        .catch(err => {
+          // Uh oh spaghettios
+          throw err
+        })
+    }, DRAIN_INTERVAL)
+  }
```

Comment:
Store global `timeoutId`. Only start `setTimeout` if `timeoutId == null`. Set `timeoutId = null` if `memoryOnlyFileQueue` is empty or on caught error.